File: benchmarks/letlam.hs
{-# LANGUAGE RankNTypes #-}

data N = Z | S N

slow Z     = 1
slow (S n) =
  let k = \x -> x (slow n)
  in k (\x->x) + k (\x->x)

main :: IO ()
main = print (slow (S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(Z))))))))))))))))))))))))))))))

File: benchmarks/letlam.js
const S = x => ["S", x];
const Z = x => ["Z"];

let slow = n => {
  switch (n[0]) {
    case "Z":
      return 1;
    case "S":
      let rec = x => x(slow(n[1]));
      return rec(x => x) + rec(x => x);
  }
};

console.log(slow(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(S(Z()))))))))))))))))))))))))))))));
